File: src/reading_plan/input/validate.py
"""Validate normalized books and settings before they reach the planner."""

from typing import TYPE_CHECKING

from reading_plan.planner_types import PLAN_MODES, WEEKDAYS


if TYPE_CHECKING:
    from typing import Literal

    from reading_plan.planner_types import Book, Settings

MIN_PROGRESS_PERCENT = 0
MAX_PROGRESS_PERCENT = 100
# Books can be prioritized 1 (highest) to 5 (lowest).
BOOK_PRIORITY_RANGE = range(1, 6)
# Books can have difficulty 1 (easiest) to 10 (hardest).
BOOK_DIFFICULTY_RANGE = range(1, 11)
VALID_WEEKDAYS = frozenset(WEEKDAYS)
VALID_DIFFICULTY_KEYS = frozenset(BOOK_DIFFICULTY_RANGE)


def check_condition(
    msg: str,
    *,
    error_type: Literal["type", "value"] = "value",
    condition: bool,
) -> None:
    """Raise an error based on error_type.

    :param msg: The error message to include if the condition is not met.
    :param error_type: The type of error to raise if the condition is not met.
    :param condition: The boolean condition to check. If False, an error is
                        raised.
    """
    if condition:
        return
    if error_type == "type":
        raise TypeError(msg)
    if error_type == "value":
        raise ValueError(msg)


def _validate_required_fields(book: Book) -> None:
    """Validate required fields and core range constraints."""
    check_condition(
        f"book_id and title are required for {book}",
        condition=bool(book.book_id and book.title),
    )

    check_condition(
        f"There are no words left to schedule for {book.book_id}",
        condition=book.remaining_words > 0,
    )

    check_condition(
        f"priority must be between 1 and 5 for {book.book_id}",
        condition=book.priority in BOOK_PRIORITY_RANGE,
    )

    check_condition(
        f"difficulty must be 1..10 for {book.book_id}",
        condition=book.difficulty in BOOK_DIFFICULTY_RANGE,
    )

    check_condition(
        f"min_blocks_per_session must be > 0 for {book.book_id}",
        condition=book.min_blocks_per_session > 0,
    )


def _validate_book_progress(book: Book) -> None:
    """Validate read-progress and words consistency values."""
    check_condition(
        f"words_full cannot be less than remaining_words for {book.book_id}",
        condition=book.words_full is None
        or book.words_full >= book.remaining_words,
    )

    check_condition(
        f"progress_percent must be between 0 and 100 for {book.book_id}",
        condition=book.progress_percent is None
        or (
            MIN_PROGRESS_PERCENT
            <= book.progress_percent
            <= MAX_PROGRESS_PERCENT
        ),
    )


def _validate_book_limits(book: Book) -> None:
    """Validate optional daily limits and blocker invariants."""
    check_condition(
        f"max_minutes_per_day must be > 0 for {book.book_id}",
        condition=book.max_minutes_per_day is None
        or book.max_minutes_per_day > 0,
    )

    check_condition(
        f"book {book.book_id} cannot block itself",
        condition=book.blocked_by is None or book.blocked_by != book.book_id,
    )


def _validate_scheduled_days(book: Book) -> None:
    """Validate book-level scheduled weekday constraints."""
    check_condition(
        f"scheduled_days is required for {book.book_id}",
        condition=bool(book.scheduled_days),
    )

    check_condition(
        f"scheduled_days must be Mon..Sun for {book.book_id}",
        condition=set(book.scheduled_days) <= VALID_WEEKDAYS,
    )


def validate_book(book: Book) -> None:
    """Validate book."""
    _validate_required_fields(book)
    _validate_book_progress(book)
    _validate_book_limits(book)
    _validate_scheduled_days(book)


def validate_settings(settings: Settings) -> None:
    """Validate settings."""
    _validate_settings_dates(settings)
    _validate_settings_minutes(settings)
    _validate_settings_positive_limits(settings)
    _validate_settings_weekday_minutes(settings)
    _validate_settings_difficulty_multiplier(settings)
    _validate_settings_plan_mode(settings)


def _validate_settings_dates(settings: Settings) -> None:
    """Validate settings date ordering."""
    check_condition(
        "end_date must be on or after start_date",
        condition=settings.start_date <= settings.end_date,
    )


def _validate_settings_minutes(settings: Settings) -> None:
    """Validate settings minute sources and quantum values."""
    check_condition(
        "Set minutes_per_day or minutes_by_weekday in settings",
        condition=bool(settings.minutes_per_day or settings.minutes_by_weekday),
    )

    check_condition(
        "time_quantum_minutes must be set to a positive integer in settings",
        condition=settings.time_quantum_minutes is not None
        and settings.time_quantum_minutes > 0,
    )


def _validate_settings_positive_limits(settings: Settings) -> None:
    """Validate positive daily session and book limits."""
    check_condition(
        "Set max_sessions_per_day and max_books_per_day to positive integers",
        condition=settings.max_sessions_per_day > 0
        and settings.max_books_per_day > 0,
    )


def _validate_settings_weekday_minutes(settings: Settings) -> None:
    """Validate weekday minute overrides when they are provided."""
    weekday_keys = set(settings.minutes_by_weekday)
    check_condition(
        "minutes_by_weekday keys must be Mon..Sun when provided",
        condition=not weekday_keys or weekday_keys == VALID_WEEKDAYS,
    )


def _validate_settings_difficulty_multiplier(settings: Settings) -> None:
    """Validate difficulty multiplier coverage."""
    difficulty_keys = set(settings.difficulty_multiplier)
    check_condition(
        "difficulty_multiplier must be empty or contain keys 1..10",
        condition=(
            not difficulty_keys or difficulty_keys == VALID_DIFFICULTY_KEYS
        ),
    )


def _validate_settings_plan_mode(settings: Settings) -> None:
    """Validate planner mode selection."""
    check_condition(
        f"plan_mode must be set to one of: {', '.join(PLAN_MODES)}",
        condition=settings.plan_mode in PLAN_MODES,
    )

File: apps/bartleby/src-tauri/src/native_planner/validate.rs
use std::collections::{BTreeSet, HashMap, HashSet};

use crate::native_planner::models::{
    Book, Settings, PLAN_MODE_FINISH_SOON, PLAN_MODE_SPREAD_OUT, WEEKDAYS,
};

const MAX_PROGRESS_PERCENT: f64 = 100.0;
const MIN_PROGRESS_PERCENT: f64 = 0.0;

struct BlockerWalk<'a> {
    by_id: HashMap<&'a str, &'a Book>,
    visited: HashSet<&'a str>,
    visiting: HashSet<&'a str>,
}

pub fn validate_blockers(books: &[Book]) -> Result<(), String> {
    let by_id = books
        .iter()
        .map(|book| (book.book_id.as_str(), book))
        .collect();
    for book in books {
        validate_blocker_reference(book, &by_id)?;
    }
    let mut blocker_walk = BlockerWalk::new(by_id);
    for book in books {
        blocker_walk.walk(book.book_id.as_str())?;
    }
    Ok(())
}

pub fn validate_book(book: &Book) -> Result<(), String> {
    if book.book_id.is_empty() || book.title.is_empty() {
        return Err(format!(
            "book_id and title are required for {}",
            book.book_id
        ));
    }
    if book.remaining_words <= 0 {
        return Err(format!(
            "There are no words left to schedule for {}",
            book.book_id
        ));
    }
    if book.priority < 1 {
        return Err(format!("priority must be >= 1 for {}", book.book_id));
    }
    if !(1..=10).contains(&book.difficulty) {
        return Err(format!("difficulty must be 1..10 for {}", book.book_id));
    }
    if book.min_blocks_per_session <= 0 {
        return Err(format!(
            "min_blocks_per_session must be > 0 for {}",
            book.book_id
        ));
    }
    if book
        .words_total
        .filter(|words_total| *words_total < book.remaining_words)
        .is_some()
    {
        return Err(format!(
            "words_total cannot be less than remaining_words for {}",
            book.book_id
        ));
    }
    if !(MIN_PROGRESS_PERCENT..=MAX_PROGRESS_PERCENT).contains(&book.progress_percent) {
        return Err(format!(
            "progress_percent must be between 0 and 100 for {}",
            book.book_id
        ));
    }
    if book
        .max_minutes_per_day
        .filter(|value| *value <= 0)
        .is_some()
    {
        return Err(format!(
            "max_minutes_per_day must be > 0 for {}",
            book.book_id
        ));
    }
    if book.blocked_by.as_deref() == Some(book.book_id.as_str()) {
        return Err(format!("book {} cannot block itself", book.book_id));
    }
    validate_scheduled_days(&book.scheduled_days, &book.book_id)
}

pub fn validate_settings(settings: &Settings) -> Result<(), String> {
    if settings.start_date > settings.end_date {
        return Err("end_date must be on or after start_date".to_string());
    }
    if settings.minutes_per_day.unwrap_or(0) <= 0 && settings.minutes_by_weekday.is_empty() {
        return Err("Set minutes_per_day or minutes_by_weekday in settings".to_string());
    }
    if settings.time_quantum_minutes <= 0 {
        return Err(
            "time_quantum_minutes must be set to a positive integer in settings".to_string(),
        );
    }
    if settings.max_sessions_per_day <= 0 || settings.max_books_per_day <= 0 {
        return Err(
            "Set max_sessions_per_day and max_books_per_day to positive integers".to_string(),
        );
    }
    let weekday_keys: BTreeSet<&str> = settings
        .minutes_by_weekday
        .keys()
        .map(String::as_str)
        .collect();
    let valid_weekdays: BTreeSet<&str> = WEEKDAYS.into_iter().collect();
    if !weekday_keys.is_subset(&valid_weekdays) {
        return Err("minutes_by_weekday keys must be Mon..Sun when provided".to_string());
    }
    let difficulty_keys: BTreeSet<i64> = settings.difficulty_multiplier.keys().copied().collect();
    let valid_difficulty_keys: BTreeSet<i64> = (1..=10).collect();
    if !difficulty_keys.is_empty() && difficulty_keys != valid_difficulty_keys {
        return Err("difficulty_multiplier must be empty or contain keys 1..10".to_string());
    }
    if settings.plan_mode != PLAN_MODE_FINISH_SOON && settings.plan_mode != PLAN_MODE_SPREAD_OUT {
        return Err(format!(
            "plan_mode must be set to one of: {PLAN_MODE_FINISH_SOON}, {PLAN_MODE_SPREAD_OUT}"
        ));
    }
    Ok(())
}

fn validate_scheduled_days(days: &BTreeSet<String>, book_id: &str) -> Result<(), String> {
    if days.is_empty() {
        return Err(format!("scheduled_days is required for {book_id}"));
    }
    let invalid_days = days
        .iter()
        .filter(|day| !WEEKDAYS.contains(&day.as_str()))
        .cloned()
        .collect::<Vec<_>>();
    if !invalid_days.is_empty() {
        return Err(format!("scheduled_days must be Mon..Sun for {book_id}"));
    }
    Ok(())
}

fn validate_blocker_reference(book: &Book, by_id: &HashMap<&str, &Book>) -> Result<(), String> {
    let Some(blocked_by) = &book.blocked_by else {
        return Ok(());
    };
    if by_id.contains_key(blocked_by.as_str()) {
        return Ok(());
    }
    Err(format!(
        "book {} is blocked by missing book_id {}",
        book.book_id, blocked_by
    ))
}

impl<'a> BlockerWalk<'a> {
    fn new(by_id: HashMap<&'a str, &'a Book>) -> Self {
        Self {
            by_id,
            visited: HashSet::new(),
            visiting: HashSet::new(),
        }
    }

    fn walk(&mut self, book_id: &'a str) -> Result<(), String> {
        match already_walked(&self.visited, book_id) {
            true => Ok(()),
            false => self.walk_unvisited(book_id),
        }
    }

    fn walk_unvisited(&mut self, book_id: &'a str) -> Result<(), String> {
        ensure_not_visiting(&mut self.visiting, book_id)?;
        walk_blocker(self, book_id)?;
        self.visiting.remove(book_id);
        self.visited.insert(book_id);
        Ok(())
    }
}

fn already_walked(visited: &HashSet<&str>, book_id: &str) -> bool {
    visited.contains(book_id)
}

fn ensure_not_visiting<'a>(
    visiting: &mut HashSet<&'a str>,
    book_id: &'a str,
) -> Result<(), String> {
    if visiting.insert(book_id) {
        return Ok(());
    }
    Err("blockers contain a cycle; remove circular dependencies".to_string())
}

fn walk_blocker<'a>(blocker_walk: &mut BlockerWalk<'a>, book_id: &'a str) -> Result<(), String> {
    let blocker = blocker_walk.by_id[book_id].blocked_by.as_deref();
    let Some(blocker) = blocker else {
        return Ok(());
    };
    blocker_walk.walk(blocker)
}
